source/develop/release-management/features/virt/engine-dedicated-cpus.md

```diff
@@ -0,0 +1,139 @@
+---
+Title: Engine Dedicated CPUs
+Category: feature
+Authors: sabusale
+---
+
+# Engine Dedicated CPUs
+
+## Summary
+
+Once choosing dedicated CPU policy, The guest vCPUs will be strictly pinned to a set of host pCPUs
+(similarly to static CPU pinning). The set of pCPUs will be chosen to match the required guest CPU topology.
+
+Engine sets the pinning automatically when starting the VM as opposed to the manual pinning that is
+done by the user and is static.
+
+## Owner
+
+-   Name: Saif Abu Saleh (sabusale)
+-   Email: sabusale@redhat.com
+
+## Responsibilities
+
+VDSM:
+
+- Makes sure that dedicated VMs vCPUs stay exclusively pinned to the given pCPU
+
+Engine:
+- Holds cached available cpus based on host capabilities and pinned CPUs
+- Schedule VM to run on best satisfying host based on the above cached data structure
+
+
+
+## Backend
+
+1. Add dedicated to CPU Pinning Policy enum
+2. Save host capabilities CPU topology in VM Dynamic
+3. Hold hostToAvailableCpu data structure:
+
+    Map Host id -> availableCpu object
+
+    availableCpu:
+
+    Map socket-id -> Physical-cpus
+
+    Socket: socket nr
+
+    Physical-cpus(List of String) e.g: 1,2,3
+
+    **Save/Update cached data structure:**    
+
+    Option 1:
+    * On updateVDSStatisticsData update the data structure with host capabilities
+    * Poll every X seconds all VMs and find which VM’s have cpu pinning and update the shared CPU pinning data structure based on VM pinned CPUs and host capabilities
+
+    Option 2:
+    * On updateVDSStatisticsData update the data structure with host capabilities
+    * On every start/stop/migrate/… VM update the shared CPU pinning data structure
+    With used pinning
+
+   **Init cached data structure (on engine start/stop):**
+   
+    Option1:
+    * Will be calculated after engine started after getting host caps and VMs cpu pinnings
+    
+    Option2:
+    * Create table in DB and save/read it from DB
+
+4. Scheduling
+    1. Start VM:
+        - Scheduling phase
+            - Reads hostToAvailableCpu from step 3
+            - Filter hosts that can't match virtual CPU topology (pCpu pinning should
+              be close to the virtual topology, e.g: 2 vCPU on same vSocket should be placed on same pSocket)
+            - Score all hosts to find best matching host with enough pCPUs 
+        - Run phase
+            - Based on Step 3 -Save/Update cached data structure
+            - Update VM CurrentCpuPinning DB field with the mapping
+
+    2. Stop/Destroy VM:
+        - Clear VM dynamic CPU pinning field
+        - Based on Step 3 -Save/Update cached data structure
+    
+3. Send vcpupin with cpusets and policy name in VM metadata
+   1. Send the policy name, available policy names: { none, pin, dedicated, siblings, isolate-threads }.
```

Comment:
siblings and isolate-threads are core related, although nothing in the feature page touches it.

Reply:
OK, I see it's TBD although mentioned here. A bit confusing.